keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md

```diff
@@ -0,0 +1,885 @@
+nitty-gritty.
+-->
+
+We need to update API server to support the newly proposed field and update PV controller to set a value of the new
+timestamp field when a volume transitions to a different phase. Also, if the feature gate is disabled the value must be
+re-set to `nil` when updating or creating a volume.
+
+The value of the field is not intended for use by any other Kubernetes components at this point and should be used only
+as a convenience feature for cluster admins. Cluster admins should be able to list and sort PersistentVolumes based on
+a timestamp which indicates when the volume transitioned to a different state. 
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a cluster admin I want to use `Retain` policy for released volumes which is safer than `Delete` and gives me more
+control over volume cleanup. However, this in turn leads to a growing number of volumes that need to be monitored
+and cleaned up. The cleanup should be performed based how long the volume was not used by any claim. I should be able
+to list all volumes that are now in `Released` phase and see when it transitioned to this phase to decide which volumes
+can be safely removed.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The new field is purely informative and should not introduce any risk.
+
+## Design Details
+
+Changes required for this KEP:
+
+* kube-apiserver
+  * extend PersistentVolumeStatus type with `LastPhaseTransitionTime` field
+  * validation should check timestamp format 
+  * validation should allow update from `nil`
+  * validation should not allow updating to a point in time before the current timestamp
+
+* kube-controller-manager / PV controller
+  * update the timestamp whenever PV controller transitions PV to a different phase
+  * remove the timestamp in `LastPhaseTransitionTime` field during volume status update when feature gate is disabled
+    * extend and use the existing function for disabling a status field: https://github.com/kubernetes/kubernetes/blob/d94261e904c90578e913c42c4d2a0fc8cb30937f/pkg/api/persistentvolumeclaim/util.go#L91
```

Comment:
I think util.go is used in the API server in PrepareForUpdate / Create. Do you really want the controller to clear the field proactively?

Reply:
Removing this from KEP for now - it does not have to be that specific probably, I'll figure out the right place later.